application/src/main/java/org/togetherjava/tjbot/commands/moderation/RejoinModerationRoleListener.java

```diff
@@ -0,0 +1,118 @@
+ * rejoined a guild.
+ * <p>
+ * Such actions are realized with roles and roles are removed upon leaving a guild, making it
+ * possible for users to otherwise bypass a mute by simply leaving and rejoining a guild. This class
+ * listens for join events and reapplies these roles in case the user is supposed to be e.g. muted
+ * still (according to the {@link ModerationActionsStore}).
+ */
+public final class RejoinModerationRoleListener implements EventReceiver {
+    private static final Logger logger =
+            LoggerFactory.getLogger(RejoinModerationRoleListener.class);
+
+    private final ModerationActionsStore actionsStore;
+    private final List<ModerationRole> moderationRoles;
+
+    /**
+     * Constructs an instance.
+     *
+     * @param actionsStore used to store actions issued by this command and to retrieve whether a
+     *        user should be e.g. muted
+     * @param config the config to use for this
+     */
+    public RejoinModerationRoleListener(@NotNull ModerationActionsStore actionsStore,
+            @NotNull Config config) {
+        this.actionsStore = actionsStore;
+
+        moderationRoles = List.of(
+                new ModerationRole("mute", ModerationAction.MUTE, ModerationAction.UNMUTE,
+                        guild -> ModerationUtils.getMutedRole(guild, config).orElseThrow()),
+                new ModerationRole("quarantine", ModerationAction.QUARANTINE,
+                        ModerationAction.UNQUARANTINE,
+                        guild -> ModerationUtils.getQuarantinedRole(guild, config).orElseThrow()));
+    }
+
+    @Override
+    public void onEvent(@NotNull GenericEvent event) {
+        if (event instanceof GuildMemberJoinEvent joinEvent) {
+            onGuildMemberJoin(joinEvent);
+        }
+    }
+
+    private void onGuildMemberJoin(@NotNull GuildMemberJoinEvent event) {
+        Member member = event.getMember();
+
+        for (ModerationRole moderationRole : moderationRoles) {
+            if (shouldApplyModerationRole(moderationRole, member)) {
+                reapplyModerationRole(moderationRole, member);
+            }
+        }
+    }
+
+    private boolean shouldApplyModerationRole(@NotNull ModerationRole moderationRole,
+            @NotNull IPermissionHolder member) {
+        Optional<ActionRecord> lastApplyAction = actionsStore.findLastActionAgainstTargetByType(
+                member.getGuild().getIdLong(), member.getIdLong(), moderationRole.applyAction);
+        if (lastApplyAction.isEmpty()) {
+            // User was never e.g. muted
+            return false;
+        }
+
+        Optional<ActionRecord> lastRevokeAction = actionsStore.findLastActionAgainstTargetByType(
+                member.getGuild().getIdLong(), member.getIdLong(), moderationRole.revokeAction);
+        if (lastRevokeAction.isEmpty()) {
+            // User was never e.g. unmuted
+            return isActionEffective(lastRevokeAction.orElseThrow());
+        }
```

Comment:
I'm extremely confused.

if there's no action

throw?

I mean, lastRevokeAction is empty? So orElseThrow throws?

Reply:
Actually, I might have done a typo during refactoring here. (This class already existed and was renamed + some extras)

Reply:
Awesome spot. fixed.